hibernate-core/src/main/java/org/hibernate/engine/internal/Cascade.java

```diff
@@ -96,6 +96,12 @@ public static void cascade(
 					if ( hasUninitializedLazyProperties && persister.getPropertyLaziness()[i] && action.performOnLazyProperty() ) {
 						LazyAttributeLoadingInterceptor interceptor = persister.getInstrumentationMetadata().extractInterceptor( parent );
 						child = interceptor.fetchAttribute( parent, propertyName );
```

Comment:
@barreiro has pointed out a flaw here. When an extra-lazy collection is set back into the parent, the collection gets loaded from the database.

If the collection is not extra-lazy, then child will already be a reference to an initialized PersistentCollection.

In other words, the flaw ony affects extra-lazy collections